tests/test_strategy_external.py

```diff
@@ -0,0 +1,134 @@
+import unittest
+from unittest.mock import MagicMock
+from parameterized import parameterized
+import json
+from d3a_interface.constants_limits import ConstSettings
+from d3a.models.area import Area
+from d3a.models.market.two_sided_pay_as_bid import TwoSidedPayAsBid
+from d3a.models.strategy.external_strategy import ExternalStrategy
+import d3a.models.market.market_redis_connection
+
+d3a.models.market.market_redis_connection.BlockingCommunicator = MagicMock
+d3a.models.market.market_redis_connection.ResettableCommunicator = MagicMock
+
+
+class TestExternalStrategy(unittest.TestCase):
+
+    def setUp(self):
+        self.area = Area(name="test_area")
+        self.parent_area = Area(name="parent_area")
+        self.test_market = TwoSidedPayAsBid(name="test_market")
+        self.parent_area._markets.markets = {1: self.test_market}
+        self.area.parent = self.parent_area
+        self.external_strategy = ExternalStrategy(self.area)
+        self.external_redis = self.external_strategy.redis
+
+    def tearDown(self):
+        ConstSettings.IAASettings.MARKET_TYPE = 1
+
+    @parameterized.expand([(2, ), (3, )])
+    def test_sub_to_external_requests_exposes_correct_channels_two_sided_market(self, market_type):
+        ConstSettings.IAASettings.MARKET_TYPE = market_type
+        self.external_strategy2 = ExternalStrategy(self.area)
+        self.external_strategy2.redis.redis_db.sub_to_multiple_channels.assert_called_once_with(
+            {
+                "parent-area/test-area/offer": self.external_strategy2.redis._offer,
+                "parent-area/test-area/delete_offer": self.external_strategy2.redis._delete_offer,
+                "parent-area/test-area/delete_bid": self.external_strategy2.redis._delete_bid,
+                "parent-area/test-area/bid": self.external_strategy2.redis._bid,
+            }
+        )
+
+    def test_sub_to_external_requests_exposes_correct_channels_one_sided_market(self):
+        self.external_redis.redis_db.sub_to_multiple_channels.assert_called_once_with(
+            {
+                "parent-area/test-area/offer": self.external_redis._offer,
+                "parent-area/test-area/delete_offer": self.external_redis._delete_offer,
+                "parent-area/test-area/accept_offer": self.external_redis._accept_offer,
+                "parent-area/test-area/offers": self.external_redis._offer_lists,
+            }
+        )
+
+    def _assert_dict_is_the_same_as_offer(self, offer_dict, offer):
+        assert offer.id == offer_dict["id"]
+        assert offer.real_id == offer_dict["real_id"]
+        assert offer.price == offer_dict["price"]
+        assert offer.energy == offer_dict["energy"]
+        assert offer.seller == offer_dict["seller"]
+
+    def test_list_offers(self):
```

Comment:
When I am reading this. I guess we dont need a list_bids functionality for the external strategy?

Reply:
Actually we do not. For two-sided market, the external device does neither list nor accept the bids, the device only manages its own bids (via placing and deleting a bid). The clearing of bids and offers is done exclusively on the markets, that's why there is no need for external connections to see the available bids and offers to the market.